### PR TITLE
*: update rustc and clippy

### LINKED ISSUE #3299: clippy warnings in master
## Bug Report

**What did you do?**

```
make dev
```

**What did you see instead?**

```
warning: casting u32 to i64 may become silently lossy if types change
   --> src/util/mpsc.rs:354:18
    |
354 |         assert!((timer.elapsed().subsec_nanos() as i64 - 100_000_000).abs() < 1_000_000);
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `i64::from(timer.elapsed().subsec_nanos())`
    |
    = note: #[warn(cast_lossless)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.207/index.html#cast_lossless

warning: casting u32 to i64 may become silently lossy if types change
   --> src/util/mpsc.rs:382:18
    |
382 |         assert!((timer.elapsed().subsec_nanos() as i64 - 100_000_000).abs() < 1_000_000);
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `i64::from(timer.elapsed().subsec_nanos())`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.207/index.html#cast_lossless
```

These warnings are introduced in https://github.com/pingcap/tikv/commit/238a2fd59430366f122d93480bcc032826107a71

### PR DESCRIPTION
**Pull request by BusyJay**
The version we are using won't exit with error code when we deny clippy warnings, so let's upgrade to the latest one.

Fix #3299 and #3300.

### CODE REVIEW

**Comment by Hoverbear** [comment]
Seems the rustfmt for this nightly is broken. :(

**Comment by Hoverbear** [comment]
Hm, I thought the download was broken, seems not.

```
root@dceeae1e30eb:/# rustup install nightly-2018-07-10
info: syncing channel updates for 'nightly-2018-07-10-x86_64-unknown-linux-gnu'
info: latest update on 2018-07-10, rust version 1.29.0-nightly (9bd8458c9 2018-07-09)
info: downloading component 'rustc'
 68.8 MiB /  68.8 MiB (100 %)  10.0 MiB/s ETA:   0 s                
info: downloading component 'rust-std'
 51.8 MiB /  51.8 MiB (100 %)  10.0 MiB/s ETA:   0 s                
info: downloading component 'cargo'
info: downloading component 'rust-docs'
info: installing component 'rustc'
info: installing component 'rust-std'
info: installing component 'cargo'
info: installing component 'rust-docs'

  nightly-2018-07-10-x86_64-unknown-linux-gnu installed - rustc 1.29.0-nightly (9bd8458c9 2018-07-09)

root@dceeae1e30eb:/# rustup component add rustfmt-preview --toolchain nightly-2018-07-10
info: downloading component 'rustfmt-preview'
info: installing component 'rustfmt-preview'
root@dceeae1e30eb:/# uname -a
Linux dceeae1e30eb 4.17.3-1-hardened #1 SMP PREEMPT Tue Jun 26 21:12:17 CEST 2018 x86_64 x86_64 x86_64 GNU/Linux
```

Hmm...

**Comment by siddontang** [comment]
CI failed

Rest LGTM

File: Makefile [hunk 1]
```diff
@@ -91,12 +91,12 @@ bench:
 	RUST_BACKTRACE=1 cargo run --release --bin bench-tikv --features "${ENABLE_FEATURES}"
 
 pre-format:
-	# unset first in case of any previous overrides
-	rustup override unset && \
-	rustup component add rustfmt-preview
+	@# unset first in case of any previous overrides
+	@if rustup override list | grep `pwd` > /dev/null; then rustup override unset; fi
+	@rustfmt --version &>/dev/null || rustup component add rustfmt-preview
```
Comment (breeswish) [inline Makefile]: rustfmt may be result in a wrong version, in this case we still need to `component add`.

File: src/coprocessor/codec/mysql/time/mod.rs [hunk 2]
```diff
@@ -58,19 +58,19 @@ const MONTH_NAMES_ABBR: &[&str] = &[
 ];
 
 #[inline]
-fn zero_time(tz: &FixedOffset) -> DateTime<FixedOffset> {
+fn zero_time(tz: FixedOffset) -> DateTime<FixedOffset> {
```
Comment (breeswish) [inline src/coprocessor/codec/mysql/time/mod.rs]: wired. Why are these & removed?

**Pull request merged**